### PR TITLE
save Selection state too in refocus()

### PR DESCRIPTION
**Pull request by jeremybmerrill**
see discussion in the other thread

### CODE REVIEW

File: extension/src/parser.js [hunk 1]
```diff
@@ -181,8 +181,27 @@ const getTargeting = ad => {
 // restore the state of the page when the extension clicks around.
 const refocus = cb => {
   const focus = document.activeElement;
+  const ranges = [];
+  for (var i = 0; i < window.getSelection().rangeCount; i++) {
+    ranges.push([
+      window.getSelection().getRangeAt(i).startContainer,
```
Comment (thejefflarson) [inline extension/src/parser.js]: Just save the output of `window.getSelection().getRangeAt(i)` and reference `startContainer` etc below, right?

File: extension/src/parser.js [hunk 2]
```diff
@@ -181,8 +181,27 @@ const getTargeting = ad => {
...
+      window.getSelection().getRangeAt(i).startOffset,
+      window.getSelection().getRangeAt(i).endContainer,
+      window.getSelection().getRangeAt(i).endOffset
+    ]);
+  }
   cb();
   if (focus) focus.focus();
+  if (ranges.length > 0) {
+    var newSelection = window.getSelection();
```
Comment (thejefflarson) [inline extension/src/parser.js]: s/var/const/

File: extension/src/parser.js [hunk 3]
```diff
@@ -181,8 +181,27 @@ const getTargeting = ad => {
...
+    newSelection.removeAllRanges();
+    for (var i = 0; i < ranges.length; i++) {
+      var range = document.createRange();
```
Comment (thejefflarson) [inline extension/src/parser.js]: s/var/const/

File: extension/src/parser.js [hunk 4]
```diff
@@ -181,8 +181,27 @@ const getTargeting = ad => {
...
+      window.getSelection().getRangeAt(i).endContainer,
+      window.getSelection().getRangeAt(i).endOffset
+    ]);
+  }
   cb();
   if (focus) focus.focus();
+  if (ranges.length > 0) {
+    var newSelection = window.getSelection();
+    newSelection.removeAllRanges();
+    for (var i = 0; i < ranges.length; i++) {
```
Comment (thejefflarson) [inline extension/src/parser.js]: You can use `ranges.forEach` here, a bit nicer but nbd.

File: extension/src/parser.js (hunk 1 above)
Comment (jeremybmerrill) [reply extension/src/parser.js]: Under some circumstances, the range object gets mutated.

File: extension/src/parser.js (hunk 2 above)
Comment (jeremybmerrill) [reply extension/src/parser.js]: can do

File: extension/src/parser.js (hunk 3 above)
Comment (jeremybmerrill) [reply extension/src/parser.js]: can do

File: extension/src/parser.js (hunk 1 above)
Comment (thejefflarson) [reply extension/src/parser.js]: K, then only do one call to getSelection, we’re perf stretched as is.

Comment (jeremybmerrill) [reply extension/src/parser.js]: good call

**Pull request merged**